From: Michal Kubecek <mkubecek@suse.cz>
Date: Fri, 2 Oct 2015 10:29:17 +0200
Subject: kABI: protect struct rt6_info changes from bsc#947321 changes
Patch-mainline: Never, kabi workaround
References: bsc#947321

Move the newly added rt6i_pmtu member into a 4-byte hole between
rt6i_metric and rt6i_idev (present on all SLE12 architectures).

Hide this member from genksyms but add a build-time check to make sure
the layout does not change.

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 include/net/ip6_fib.h | 5 +++--
 net/ipv6/ip6_fib.c    | 6 ++++++
 2 files changed, 9 insertions(+), 2 deletions(-)

diff --git a/include/net/ip6_fib.h b/include/net/ip6_fib.h
index 63b6d49cb7b1..464a32d09c3b 100644
--- a/include/net/ip6_fib.h
+++ b/include/net/ip6_fib.h
@@ -115,13 +115,14 @@ struct rt6_info {
 	struct rt6key			rt6i_src;
 	struct rt6key			rt6i_prefsrc;
 	u32				rt6i_metric;
+#ifndef __GENKSYMS__
+	u32				rt6i_pmtu;
+#endif
 
 	struct inet6_dev		*rt6i_idev;
 	unsigned long			_rt6i_peer;
 
 	u32				rt6i_genid;
-
-	u32				rt6i_pmtu;
 	/* more non-fragment space at head required */
 	unsigned short			rt6i_nfheader_len;
 
diff --git a/net/ipv6/ip6_fib.c b/net/ipv6/ip6_fib.c
index 6e900cac429b..99bc176f90d4 100644
--- a/net/ipv6/ip6_fib.c
+++ b/net/ipv6/ip6_fib.c
@@ -1835,6 +1835,12 @@ int __init fib6_init(void)
 {
 	int ret = -ENOMEM;
 
+	/* check kABI correctness of bsc#947321 changes */
+	BUILD_BUG_ON(offsetof(struct rt6_info, rt6i_idev) !=
+		     offsetof(struct rt6_info, rt6i_metric) +
+		     sizeof(((struct rt6_info *)NULL)->rt6i_metric) +
+		     sizeof(((struct rt6_info *)NULL)->rt6i_pmtu));
+
 	fib6_node_kmem = kmem_cache_create("fib6_nodes",
 					   sizeof(struct fib6_node),
 					   0, SLAB_HWCACHE_ALIGN,
-- 
2.5.3

